### ISSUE TITLE
[RFC] : Dynamically Quantized Matrix Multiplication with  8-bit Support

### ISSUE THREAD

**Issue by nikhil-arm** [pytorch/pytorch#149500]
This RFC proposes extending PyTorch's dynamic quantized matmul capabilities to support both 4-bit and 8-bit quantized matrix multiplication. Two potential integration paths are presented below.

---

## Approach 1: Separate 8-Bit Operators

This approach introduces new dedicated operators for 8-bit dynamic quantization. These operators closely mirror the existing 4-bit operators in design and functionality.

### 1. `torch.ops.aten._dyn_quant_pack_8bit_weight`

This operation packs the 8-bit quantized weights, scales, and bias for a Linear layer into a compact format using 8-bit symmetric quantization.

#### **Parameters:**

- **`weight`** (`Tensor`): The original weights of the Linear layer.
- **`scales_and_zeros`** (`Tensor`): A tensor containing the quantization scales (and possibly zero-points) for each group. The tensor has the shape `[num_groups]`.
- **`bias`** (`Tensor`, optional): The bias tensor for the Linear layer.
- **`groupsize`** (`int`): The number of channels per group. Must be a multiple of 32 or equal to `in_features`.
- **`in_features`** (`int`): The number of input features.
- **`out_features`** (`int`): The number of output features.

#### **Returns:**
A tensor representing the packed 8-bit weights and associated quantization parameters, ready to be used in the matrix multiplication.

---

### 2. `torch.ops.aten._dyn_quant_matmul_8bit`

This operation performs matrix multiplication using the 8-bit quantized weights.

#### **Parameters:**

- **`input`** (`Tensor`): The input tensor for matrix multiplication, typically with shape `[batch_size, in_features]`.
- **`packed_weights`** (`Tensor`): The packed 8-bit weights, as produced by `torch.ops.aten._dyn_quant_pack_8bit_weight`.
- **`groupsize`** (`int`): The number of channels per group.
- **`in_features`** (`int`): The number of input features.
- **`out_features`** (`int`): The number of output features.

#### **Returns:**
A tensor containing the result of the matrix multiplication with shape `[batch_size, out_features]`.

---

## Approach 2: Unified n-Bit Operators

In this approach, we generalize the current 4-bit operators ```_dyn_quant_matmul_4bit``` to support multiple bit-widths (currently 4 and 8). A new parameter (e.g., `bit_width` or `n`) is introduced to control the precision, and the operators dispatch to the appropriate 4-bit or 8-bit kernels accordingly.

### 1. `torch.ops.aten._dyn_quant_pack_weight`

This unified packing operation now supports both 4-bit and 8-bit quantization by accepting a `bit_width` parameter.

#### **Parameters:**

- **`weight`** (`Tensor`): The original weights of the Linear layer.
- **`scales_and_zeros`** (`Tensor`): A tensor containing quantization scales (and zero-points if necessary) for each group.
- **`bias`** (`Tensor`, optional): The bias tensor for the Linear layer.
- **`groupsize`** (`int`): The number of channels per group.
- **`in_features`** (`int`): The number of input features.
- **`out_features`** (`int`): The number of output features.
- **`bit_width`** (`int`): The precision of quantization. Accepted values are 4 or 8. (Defaults to 4 if not specified.)

#### **Returns:**
A tensor representing the packed weights and quantization parameters in the chosen bit precision.

---

### 2. `torch.ops.aten._dyn_quant_matmul`

This unified matrix multiplication operation supports both 4-bit and 8-bit quantized weights.

#### **Parameters:**

- **`input`** (`Tensor`): The input tensor for matrix multiplication, typically with shape `[batch_size, in_features]`.
- **`packed_weights`** (`Tensor`): The packed weights as produced by `torch.ops.aten._dyn_quant_pack_weight`.
- **`groupsize`** (`int`): The number of channels per group.
- **`in_features`** (`int`): The number of input features.
- **`out_features`** (`int`): The number of output features.
- **`bit_width`** (`int`): Indicates the bit-width of the quantization (4 or 8) and controls which kernel is used for computation.

#### **Returns:**
A tensor containing the result of the matrix multiplication with shape `[batch_size, out_features]`.

---

### Using Separate 8-Bit Operators:

```python
# 8-bit weight packing
packed_weights_8bit = torch.ops.aten._dyn_quant_pack_8bit_weight(
    weight, scales_and_zeros, bias, groupsize, in_features, out_features
)

# 8-bit matrix multiplication
output = torch.ops.aten._dyn_quant_matmul_8bit(
    input, packed_weights_8bit, groupsize, in_features, out_features
)


# Packing weights with a specified bit width (4 or 8)
packed_weights = torch.ops.aten._dyn_quant_pack_weight(
    weight, scales_and_zeros, bias, groupsize, in_features, out_features, bit_width=8
)

# Matrix multiplication using the unified operator
output = torch.ops.aten._dyn_quant_matmul(
    input, packed_weights, groupsize, in_features, out_features, bit_width=8
)

**Comment by digantdesai** [comment]
@nikhil-arm Have you looked at `aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp` and if that can be extended?

**Comment by nikhil-arm** [comment]
Thanks @digantdesai for the inputs, @fadara01 did some research on potential issues with this path. @fadara01 can you please list them for discussion.

**Comment by fadara01** [comment]
The main issue with `aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp` is that it suffers from high overhead due to `__torch_function__` mode.

`torch.ao.quantization.quantize_dynamic` packs the weights into a subclass of [LinearPackedParamsBase](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/PackedParams.h#L6) (which extends the `torch::jit::CustomClassHolder`).
At runtime (i.e. on every `model.forward` call for every layer), there is a check to see if the `LinearPackedParamsBase` object has a `__torch_function__` attribute ([has_torch_function_attr()](https://github.com/pytorch/pytorch/blob/main/torch/csrc/utils/disable_torch_function.cpp#L310)) and that's the overhead comes from.

To give an idea about the magnitude of the overhead, for the benchmark below:
running:
`OMP_NUM_THREADS=1 python simple_dynamic_linear.py --M 1 --K 1 --N 1 --iters 100000`
is **at least 20us** slower than:
`OMP_NUM_THREADS=1 python simple_dynamic_linear.py --M 1 --K 1 --N 1 --iters 100000 --disable_torch_function_mode` (where the benchmark is wrapped under the `with torch._C.DisableTorchFunction()` context manager)

For a real workload like `bert-large`, we noticed that tuning off `__torch_function__` mode yields an 11% speedup. 

```
import torch
import time
import argparse
import torch.ao.quantization
import numpy as np
import random

seed = 42
torch.manual_seed(seed)
random.seed(seed)
np.random.seed(seed)

class LinearNet(torch.nn.Module):
   def __init__(self, K, N):
      super().__init__()
      self.fc1 =  torch.nn.Linear(K, N, bias=False)

   def forward(self, x):
      x = self.fc1(x)
      return x

 
def bench(args):
   m = int(args.M)
   k = int(args.K)
   n = int(args.N)

   with torch.no_grad():

      example_inputs = torch.randn(size=(m, k))

      model = LinearNet(k, n).eval()
      model = torch.ao.quantization.quantize_dynamic(
               model,
               {torch.nn.Linear},
               dtype=torch.qint8)

      # warmup   
      for _ in range(10):
         model(example_inputs)
      times = []
      
      for i in range(args.iters):
         example_inputs = torch.randn(size=(m, k))

         s = time.time_ns()
         model(example_inputs)
         times.append((time.time_ns() - s) / 1e3)
      
   
   print(f"Mean times = {sum(times) / len(times)} us")
   print(f"Min times = {min(times)} us")
 
if __name__ == "__main__":
    
    
   parser = argparse.ArgumentParser()

   # Define optional flags with short aliases
   parser.add_argument("--M", type=int, help="M", default=1024)
   parser.add_argument("--K", type=int, help="K", default=1024)
   parser.add_argument("--N", type=int, help="N", default=1024)
   parser.add_argument("--seed", type=int, help="seed", default=42)
   parser.add_argument("--iters", type=int, help="benchmark iters", default=1000)
   parser.add_argument("--disable_torch_function_mode", action="store_true", default=False)
   args = parser.parse_args()
   if args.disable_torch_function_mode:
      with torch._C.DisableTorchFunction():
         bench(args)
   else:
      bench(args)
```

**Comment by nikhil-arm** [comment]
@digantdesai can you please review this?

**Comment by digantdesai** [comment]
> For a real workload like bert-large, we noticed that tuning off __torch_function__ mode yields an 11% speedup compared to the default path.

Thanks for the details. `PyObject_FastGetAttrString` is the main overhead?

The reason I suggested that because we can get eager support and speed up existing users of the 8b dqlinear op. This opens up yet-another-special-path.

That said ignoring this for a minute.

For this RFC, I like approach 2 i.e. `Unified n-Bit Operators Approach`

Orthogonal to this, extending aten op set is OK but any thoughts on doing this in torchao? IMO that is cleaner, and we already have unified 1b to 8b Linear op with various quant schema, and it works well with the `quantize_` APIs.

**Comment by fadara01** [comment]
> Thanks for the details. PyObject_FastGetAttrString is the main overhead?
@digantdesai yes that's correct.
Mainly from `tp->tpgetattro` [here](https://github.com/pytorch/pytorch/blob/main/torch/csrc/utils/python_strings.h#L118) which is implemented [here](https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/api/object.h#L77)
